### PR TITLE
[wip] Separate Test and Fit

### PR DESCRIPTION
**Pull request by tullie**
## What does this PR do?
Address some of the discussion from issue #1195. On the public API it changes:
```
trainer = Trainer(args)
trainer.fit(model)
trainer.test()
```
To:
```
Trainer(args).fit(model)
Evaluator(args).test(model)
```
To achieve this - internally there's a large refactor that begins to decouple training and testing code. Notably:
- Adds a shared LightingConfig class that's used for Trainer and Evaluator config options.
- Adds shared mixins such as (SlurmMixin, LoopRunnerMixin and InitializationMixin)
- Refactors fit and run_pretrain_loop functions to use callables
- Moved shared functions (num_gpus and data_parallel) to initialization
- Moves PatchDataLoader to data_loader mixin
- Refactors CLI code to work with LightningConfig and be slightly clearer

Above I say "begins to decouple" because there's still so much internal refactoring to do. I really did the least possible amount to be able to achieve this API in anticipation for a 1.0 release. Ideally, we'd be able to get rid of concepts like an OptimizerMixin from the Evaluator, however, it's all too tangled up atm. In the future i'd propose replacing the mixin architecture with something that encourages encapsulation between the components.

In regards to backwards compatibility, i've aimed to fully support the old API until version 0.9.0.

## TODO in this PR 
- Add changelog
- Test on TPU and fix this line torch_xla.core.xla_model.rendezvous("pl.Tester.run_pretrain_routine")
- Update deprecation since version numbers

### CODE REVIEW

**Comment by Borda** [comment]
btw, what about Trainer & Examinator? :rabbit:

**Comment by Borda** [comment]
I like the idea of splitting, but not sure about advantage of `LightingConfig`
does it also mean the the `.test(...)` is removed from trainer?

btw, what about Trainer & Examinator? :rabbit: 
probably it would be cleaner to make it after #2073

File: pytorch_lightning/trainer/initialization_mixin.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+        # Transfer params
+        self.num_nodes = config.num_nodes
+        # Backward compatibility, TODO: remove in v0.8.0
+        if config.nb_gpu_nodes is not None:
+            rank_zero_warn(
+                "Argument `nb_gpu_nodes` has renamed to `num_nodes` since v0.5.0"
+                " and this method will be removed in v0.8.0",
+                DeprecationWarning,
+            )
```
Comment (Borda) [inline pytorch_lightning/trainer/initialization_mixin.py:53]: if we have now the `LightningConfig` this warning and deprecation shall be called there...

File: pytorch_lightning/trainer/lightning_config.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
+from typing import Union, Optional, List, Dict, Iterable
+from dataclasses import dataclass, fields
+
+from pytorch_lightning.loggers import LightningLoggerBase
+from pytorch_lightning.callbacks import ModelCheckpoint, EarlyStopping, Callback
+from pytorch_lightning.profiler import BaseProfiler
+
+
+@dataclass
+class LightningConfig:
```
Comment (Borda) [inline pytorch_lightning/trainer/lightning_config.py:10]: this may have a disadvantage that all the arguments won't be listed in `Trainer` docs page and so a user needs to search for it also it drops IDE support of whispering possible arguments

Comment (Borda) [inline pytorch_lightning/trainer/lightning_config.py:10]: this may have a disadvantage that all the arguments won't be listed in Trainer docs page and so a user needs to search for it also it drops IDE support of whispering possible arguments

File: pytorch_lightning/trainer/initialization_mixin.py (hunk 1 above)
Comment (Borda) [inline pytorch_lightning/trainer/initialization_mixin.py:53]: if we have now the `LightningConfig` this warning and deprecation shall be called there...

**Comment by tullie** [comment]
Yeah to be clear the goal of this PR wasn't to change the config it just turned out that we needed a good way to share it between the two classes so I don't necessarily think it's an advantage either. Instead of using kwargs we could have a positional config argument of type LightingConfig (but this wouldn't be backwards compatible) or we could just keep the arg documentation. Both of these options would handle the IDE hinting. What do you think? 

.test(...) will be removed after deprecation phase, yes!

File: pytorch_lightning/trainer/initialization_mixin.py (hunk 1 above)
Comment (tullie) [reply pytorch_lightning/trainer/initialization_mixin.py:53]: Good call, will add this in the next pass

**Comment by tullie** [comment]
Anyone know how I can fix the DocTestFailure? How can I reproduce that test locally?

**Comment by awaelchli** [comment]
yes it means The trainer args have changed (either the ordering, defaults or some got added or removed)

File: tests/trainer/test_trainer_cli.py [hunk 3]
```diff
@@ -111,28 +98,3 @@ def test_argparse_args_parsing(cli_args, expected):
...
-    sys.version_info < (3, 7),
-    reason="signature inspection while mocking is not working in Python < 3.7 despite autospec"
-)
-@pytest.mark.parametrize(['cli_args', 'extra_args'], [
-    pytest.param({}, {}),
-    pytest.param({'logger': False}, {}),
-    pytest.param({'logger': False}, {'logger': True}),
-    pytest.param({'logger': False}, {'checkpoint_callback': True}),
-])
-def test_init_from_argparse_args(cli_args, extra_args):
```
Comment (awaelchli) [inline tests/trainer/test_trainer_cli.py:126]: will this test get removed?

**Comment by reactivetype** [comment]
Please consider this related issue https://github.com/PyTorchLightning/pytorch-lightning/issues/1694
It was closed but I think the issue is not yet solved

File: pytorch_lightning/trainer/lightning_config.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/trainer/lightning_config.py:10]: I like the idea of having a lightning config. I think we could maybe change it that the trainer still has all the arguments, internally wrap it to a config holding all the internal state variables as well as configuration and just make the trainer have properties for the most relevant attributes. 

This would also have the advantage, that you could possibly save and resume the whole trainer state at every moment

File: pytorch_lightning/trainer/data_loading.py [hunk 4]
```diff
@@ -310,3 +310,25 @@ def determine_data_use_amount(self, train_percent_check: float, val_percent_chec
             self.train_percent_check = overfit_pct
             self.val_percent_check = overfit_pct
             self.test_percent_check = overfit_pct
+
+
+class PatchDataLoader(object):
```
Comment (justusschock) [inline pytorch_lightning/trainer/data_loading.py:315]: Why Is this needed now but wasn't before?

File: tests/callbacks/test_callbacks.py [hunk 5]
```diff
@@ -4,118 +4,117 @@
...
 import tests.base.utils as tutils
 from pytorch_lightning import Callback
-from pytorch_lightning import Trainer, LightningModule
+from pytorch_lightning import Trainer, Evaluator, LightningModule
 from pytorch_lightning.callbacks import EarlyStopping, ModelCheckpoint
 from pytorch_lightning.loggers import TensorBoardLogger
 from tests.base import EvalModelTemplate
 
 
+class TestCallback(Callback):
```
Comment (justusschock) [inline tests/callbacks/test_callbacks.py:13]: Will we split callbacks as well or will we have a single class that has to be attached to both, Trainer and Evaluator?

**Comment by Borda** [comment]
> Anyone know how I can fix the DocTestFailure? How can I reproduce that test locally?

For local tests pls check the CircleCI config, there are the exact commands for pytest and checking/building the docs

**Comment by Borda** [comment]
> I'm not sure, but I think I'd delay this after 0.8 and make a whole refactoring for 0.9

most likely there won't be 0.9, on the other hand it would make sense do 0.8 this week, then this split together with HyperTuner the week after...

> IMO we should try to have breaking chances as infrequently as possible and rather have one huge break than multiple small ones.

This is a good point :]

**Comment by justusschock** [comment]
> you mean something similar as we wrap hyperparameters in the model to a single workspace?

yes, exactly :)

File: pytorch_lightning/trainer/lightning_config.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/lightning_config.py:10]: i also like the config... much easier to keep track of args and share among classes.
i do prefer allowing the init to be:
```
Trainer(arg1, arg2, etc...)
```

Instead of
```
Trainer(config)
```

Internally we should wrap it in the config object. Now, in theory people could use the config directly if they wanted to:
```
Trainer(**config)
```

**Comment by williamFalcon** [comment]
> Instead of using kwargs we could have a positional config argument of type LightingConfig (but this wouldn't be backwards compatible) or we could just keep the arg documentation in Trainer. Both of these options would handle the IDE hinting. What do you think?
> 
> .test(...) will be removed after deprecation phase, yes!

what problem is config trying to solve? If it's not duplicating args between Trainer and Evaluator then we should just use inheritance and have Trainer and Evaluator inherit from a superclass that lists all config stuff.

But otherwise, i think it's important to directly have all the typehinting and directly using args
```python
Trainer(arg1=1...)
```

File: pytorch_lightning/trainer/data_loading.py (hunk 4 above)
Comment (tullie) [reply pytorch_lightning/trainer/data_loading.py:315]: It's just moved from the Trainer file

File: tests/callbacks/test_callbacks.py (hunk 5 above)
Comment (tullie) [reply tests/callbacks/test_callbacks.py:13]: At the moment it's a single class that has to be attached to both. Having evaluator specific callbacks in the future might make sense though.

File: tests/trainer/test_trainer_cli.py (hunk 3 above)
Comment (tullie) [reply tests/trainer/test_trainer_cli.py:126]: Yeah I moved the init_from_argparse function completely so I don't think there's anything to test.

**Comment by tullie** [comment]
> Please consider this related issue #1694
> It was closed but I think the issue is not yet solved

Yeah we should reopen that issue. It's a good feature request. That should go in a separate PR then this though as it's not specifically related to the class separation.

> Just some questions...
> 
> I'm not sure, but I think I'd delay this after 0.8 and make a whole refactoring for 0.9 that also includes `tuner` and making the training-loop stuff a bit more modular. IMO we should try to have breaking chances as infrequently as possible and rather have one huge break than multiple small ones.

Yeah we can delay this to 0.9. I was planning to try and incorporate some follow up refactors in following PRs so putting them all in 0.9 makes sense.

> what problem is config trying to solve? If it's not duplicating args between Trainer and Evaluator then we should just use inheritance and have Trainer and Evaluator inherit from a superclass that lists all config stuff.
> 
> But otherwise, i think it's important to directly have all the typehinting and directly using args

The way this PR is framed it isn't duplicating args because evaluator is still quite coupled with the trainer. I'm hoping to keep refactoring and separate them more though. This would potentially include separating their configs. E.g. evaluator doesn't need a gradient_accumulator arg. How does the IDE typehinting work exactly? If the superclass has typed args will it show them when hovering over the subclass or something?

File: tests/trainer/test_trainer_cli.py (hunk 3 above)
Comment (awaelchli) [reply tests/trainer/test_trainer_cli.py:126]: where did it get moved? sry but I don't get it. 
These tests were part of a bugfix and are essential to make sure it works.

Comment (tullie) [reply tests/trainer/test_trainer_cli.py:126]: Ahh sorry, you're totally right, this test shouldn't be removed. I was getting it mixed up with the one I removed above. Thanks for catching this!
